lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp

```diff
@@ -0,0 +1,345 @@
+
+bool
+CheckReachability(Graph* graph, const GNode& source) {
+  katana::InsertBag<GNode> current_bag;
+  katana::InsertBag<GNode> next_bag;
+
+  current_bag.push(source);
+  graph->GetData<NodeCount>(source) = 1;
+
+  while (true) {
+    if (current_bag.begin() == current_bag.end())
+      break;
+
+    katana::do_all(
+        katana::iterate(current_bag),
+        [&](GNode n) {
+          for (auto edge : graph->edges(n)) {
+            auto dest = *(graph->GetEdgeDest(edge));
+            if (graph->GetData<NodeCount>(dest) == 0) {
+              graph->GetData<NodeCount>(dest) = 1;
+              next_bag.push(dest);
+            }
+          }
+        },
+        katana::steal());
+
+    current_bag.clear();
+    std::swap(current_bag, next_bag);
+  }
+
+  if (graph->GetData<NodeCount>(reportNode) == 0) {
+    return false;
+  }
+
+  katana::do_all(katana::iterate(*graph), [&graph](GNode n) {
+    graph->GetData<NodeCount>(n) = 0;
+  });
+
+  return true;
+}
+
+//delta stepping implementation for finding a shortest path from source to report node
+template <typename Item, typename OBIMTy, typename PushWrap, typename EdgeRange>
+void
+DeltaStepAlgo(
+    Graph* graph, const GNode& source, const PushWrap& pushWrap,
+    const EdgeRange& edgeRange,
+    katana::InsertBag<std::pair<uint32_t, Path*>>* bag,
+    katana::InsertBag<Path*>* path_pointers) {
+  //! [reducible for self-defined stats]
+  katana::GAccumulator<size_t> bad_work;
+  //! [reducible for self-defined stats]
+  katana::GAccumulator<size_t> wl_empty_work;
+
+  graph->GetData<NodeCount>(source) = 1;
+
+  katana::InsertBag<Item> init_bag;
+
+  Path* path = new Path();
+  path->last = NULL;
+  path_pointers->push(path);
+
+  pushWrap(init_bag, source, 0, path, "parallel");
+
+  katana::for_each(
+      katana::iterate(init_bag),
+      [&](const Item& item, auto& ctx) {
+        for (auto ii : edgeRange(item)) {
+          GNode dst = *(graph->GetEdgeDest(ii));
+          auto& ddata_count = graph->GetData<NodeCount>(dst);
+          auto& ddata_max = graph->GetData<NodeMax>(dst);
+
+          Distance ew = graph->GetEdgeData<EdgeWeight>(ii);
+          const Distance new_dist = item.distance + ew;
+
+          if ((ddata_count >= numPaths) && (ddata_max <= new_dist))
+            continue;
+
+          Path* path;
+          path = new Path();
+          path->parent = item.src;
+          path->last = item.path;
+          path_pointers->push(path);
+
+          if (ddata_count < numPaths) {
+            katana::atomicAdd<uint32_t>(ddata_count, (uint32_t)1);
+            katana::atomicMax<uint32_t>(ddata_max, new_dist);
+          }
+
+          if (dst == reportNode)
+            bag->push(std::make_pair(new_dist, path));
+
+          if ((graph->GetData<NodeCount>(reportNode) < numPaths) ||
+              ((graph->GetData<NodeCount>(reportNode) >= numPaths) &&
+               (graph->GetData<NodeMax>(reportNode) > new_dist))) {
```

Comment:
There's another way to terminate. `for_each` has support for a parallel break/exit. 
So a more clear way to write this is:
```
for_each(..., [&](const Item& item, auto& ctx) {
for (auto ii : edgeRange(item)) {
  GNode dst = *(graph->GetEdgeDest(ii));
  ...
  if (dst == reportNode) {
    ... 
    bool terminate = ...;
    if (terminate) {
      ctx.breakLoop();
      return;
    }
  }
  ...
  pushWrap(...);
}
}, katana::parallel_break(), ...);
```

However, I'm not sure if this would be more performant than the non-break option above. So please benchmark both before you choose one.

Reply:
Yes, both yield the same results.

Reply:
but I am not trying to check for termination. I do admit that I should have written it in a better way, like your first suggestion.

Reply:
Why not? It's safe to terminate as soon as you have the required number of paths for the report node.

Reply:
the issue is that there is still a chance that I can a get a shorter path later on. So I need to run it until all items in the worklist are processed.

Reply:
Makes sense. Yes, it's not a termination condition. I get it now.